### PR TITLE
ceph: Order CephCluster false condition before true Ready condition

### LINKED ISSUE #6977: Rancher CephCluster state shows error while ready
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
Rancher CephCluster resource details show confusing status.
The cluster is up and running fine and there are no errors further, but the 'error' state can be very confusing.

Screenshot: 

Screenshot 2021-01-14 at 19 16 28

**Environment**:

* Rook version (use `rook version` inside of a Rook Pod): 1.5.5

### PR DESCRIPTION
**Pull request by ron1**
**Description of your changes:**
Order CephCluster false condition before its corresponding true Ready condition.

**Which issue is resolved by this Pull Request:**
Resolves #6977

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Review by travisn — COMMENTED** [review]
LGTM, but still pending discussion in #6977

**Comment by travisn** [comment]
There is an issue updating the conditions with this change. When testing, I see this error in the operator log:
```
2021-02-09 19:05:33.975790 I | ceph-cluster-controller: done reconciling ceph cluster in namespace "rook-ceph"
2021-02-09 19:05:33.998007 E | op-config: failed to update cluster condition to {Type:Ready Status:True Reason:ClusterCreated Message:Cluster created successfully LastHeartbeatTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:0001-01-01 00:00:00 +0000 UTC}. Operation cannot be fulfilled on cephclusters.ceph.rook.io "my-cluster": the object has been modified; please apply your changes to the latest version and try again
```

Not sure it's related to this PR, but updating the conditions needs a deeper look...

**Comment by ron1** [comment]
@travisn I pushed a new version of the patch that simplifies the code by eliminating package variables and removing the unused ErrorMapping() function. I think this new version should fix the error reported during testing. At this point, I do not have an environment to functional test the patch. Any chance this might be backported in time for the 1.5.7 release?

File: pkg/operator/ceph/config/conditions.go
Comment (travisn) [inline pkg/operator/ceph/config/conditions.go:57]: We still should log an error so we know the conditions were not updated as expected.

File: pkg/operator/ceph/config/conditions.go
Comment (travisn) [inline pkg/operator/ceph/config/conditions.go:53]: Let's log the error

File: pkg/operator/ceph/config/conditions.go
Comment (travisn) [inline pkg/operator/ceph/config/conditions.go:50]: This code would really be improved if we computed all the conditions at once and had a single update to the CephCluster CR. Currently it involves multiple updates. But this pattern already existed, so we can keep it for now instead of blocking this PR.

File: pkg/operator/ceph/config/conditions.go
Comment (travisn) [inline pkg/operator/ceph/config/conditions.go:197]: In my testing I never see this "checking if updates are needed" on the `Progressing` condition. It's always "Cluster is creating".

**Review by travisn — CHANGES_REQUESTED** [review]
In my testing I don't see the `Ready` condition updated on later reconciles. For example, these are the conditions are creating the cluster, then restarting the operator and waiting for the reconcile to complete:

```yaml
    conditions:
    - lastHeartbeatTime: "2021-02-11T18:31:09Z"
      lastTransitionTime: "2021-02-11T18:21:48Z"
      message: Cluster progression is completed
      reason: ProgressingCompleted
      status: "False"
      type: Progressing
    - lastHeartbeatTime: "2021-02-11T18:22:31Z"
      lastTransitionTime: "2021-02-11T18:22:31Z"
      message: Cluster created successfully
      reason: ClusterCreated
      status: "True"
      type: Ready
```
If we could add the errors back to the log, I suspect we'll see the same error as before.

**Review by travisn — APPROVED** [review]

File: pkg/operator/ceph/config/conditions.go
Comment (ron1) [reply pkg/operator/ceph/config/conditions.go:57]: The error is already being logged in `getCluster()`. Since that is an internal function, I did not think it needed to be logged again.

File: pkg/operator/ceph/config/conditions.go
Comment (ron1) [reply pkg/operator/ceph/config/conditions.go:200]: Yes, I was hesitant to change any existing behavior here.

File: pkg/operator/ceph/config/conditions.go
Comment (travisn) [reply pkg/operator/ceph/config/conditions.go:57]: Ah, I overlooked that. So we will have double logging on this error since I was trigger happy with the merge. The pattern more common in rook would be to only log at the point where the error is ignored, so it's clear that it's intentionally ignored.

**Pull request merged**
